Prepare release for PWA Kit 3.14.0-preview.0

shethj opened the pull request:
<!--- Provide a short summary of your changes in the Title field above -->

# Description

Prepare PWA Kit monorepo for v3.14.0-preview.0 release. This release includes all code that was part of `develop` as on Oct 24, 2025 which was the feature freeze date for v3.14.0 release.

This PR also pulls in following PRs for product bundles:
- #3418 
- #3443 
- #3445 

This PR bumps versions for monorepo pacakges as follows:
```
- @salesforce/commerce-sdk-react => 4.2.0-preview.0
 - @salesforce/pwa-kit-create-app => 3.14.0-preview.0
 - @salesforce/pwa-kit-dev => 3.14.0-preview.0
 - @salesforce/pwa-kit-mcp => 0.5.0-dev # Not to be released
 - @salesforce/pwa-kit-react-sdk => 3.14.0-preview.0
 - @salesforce/pwa-kit-runtime => 3.14.0-preview.0
 - @salesforce/retail-react-app => 8.2.0-preview.0
```
<!--- A longer summary of your changes, including: a description of the issue that you’re addressing, a list of required dependencies (if applicable), and any other relevant context. -->

# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- (change1)

# How to Test-Drive This PR

- (step1)

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

## Accessibility Compliance

You must check off all items in **one** of the follow two lists:

- [ ] There are no changes to UI

_or..._

- [ ] Changes were tested with a Screen Reader (iOS VoiceOver or Android Talkback) and had no issues
- [ ] Changes comply with [WCAG 2.0 guidelines levels A and AA](https://www.wuhcag.com/wcag-checklist/)
- [ ] Changes to common UI patterns and interactions comply with [WAI-ARIA best practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## Localization

- [ ] Changes include a UI text update in the Retail React App (which requires translation)

shethj committed:
Bump versions for 3.14.0-preview.0

cc-prodsec commented:
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-critical.svg) Critical | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-high.svg) High | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-medium.svg) Medium | ![](https://res.cloudinary.com/snyk/image/upload/r-d/scm-platform/snyk-pull-requests/severity-low.svg) Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Licenses| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/8642001b-aaa2-4939-9f36-a89c845506d0/license?refs=6a64b1b3650d7ae78f60fc30dd2b9c67e9cf46c8&source=prChecksComment) |
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/cc-mobify-sdk/pr-checks/8642001b-aaa2-4939-9f36-a89c845506d0?refs=6a64b1b3650d7ae78f60fc30dd2b9c67e9cf46c8&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

sf-shikhar-prasoon committed:
Add support for Rule Based Promotions for Choice of Bonus Products (#3418)

* @W-18957985- Product Search Integration & Rule-Based Detection (#3396)

* 1. use search endpoint

* 2 add rule based detection

* 3. update discovery logic

* add tests, update bundle size limit, update changelog

* @W-19980361 Rule based bonus products- Work Item 2- update display logic (#3408)

* 1. use search endpoint

* 2 add rule based detection

* 3. update discovery logic

* cleanup. remove unused code

* add tests, update bundle size limit, update changelog

* fix header counts

* fix flasing, re-rendering

* pr comments + lint

* @W-19980361 fix 2 bugs- selection count & flashing due to re-renders (#3421)

* fix header counts

* fix flasing, re-rendering

* pr comments + lint

* add param &refine=pmpt=bonus for promotion type

* pr comments

* use commerce sdk

* add const- limit=50  |  replace harded Search Param values

* remove function breakdown change

* update changelog

* refactor. re-use

* fix: prevent promotion message flashing when changing product variant… (#3428)

* fix tests

* Rule-Based Bonus Product Promotions Enhancement

This PR enhances the existing rule-based bonus product promotions feature by adding proper variant eligibility checking. Previously, rule-based promotions could not accurately determine which specific product variants qualified for a promotion. This enhancement introduces a qualifying products lookup system that uses the product search API to fetch and validate eligible variants, ensuring that only products that truly qualify can trigger bonus product selection UI and functionality.

Key Changes:

1. New Hook: useRuleBasedQualifyingProducts
   - Added to app/hooks/use-rule-based-bonus-products.js
   - Fetches qualifying products for rule-based promotions using productSearch API
   - Uses refinement parameters: pmid=${promotionId} and pmpt=qualifying
   - Returns a Set of qualifying product IDs for efficient lookup
   - Supports variant eligibility checking by examining both variant and master product IDs

2. Enhanced Promotion Eligibility Logic
   - Updated getPromotionIdsForProduct() in app/utils/bonus-product/common.js
   - Distinguishes between list-based and rule-based promotions
   - Added fallback logic to check master product IDs when variants don't directly qualify
   - Filters promotion IDs to only return those the product actually qualifies for

3. Qualifying Products Map Integration
   - Added ruleBasedQualifyingProductsMap parameter throughout the bonus product utility chain
   - Updated useBasketProductsWithPromotions() hook to fetch and return qualifying products map
   - Propagated the map through all relevant functions and components

4. Updated Function Signatures
   - isAutomaticPromotion()
   - shouldShowBonusProductSelection()
   - getBonusProductsForSpecificCartItem()
   - getBonusProductsInCartForProduct()
   - getAvailableBonusItemsForProduct()
   - getRemainingAvailableBonusProductsForProduct()

5. Component Updates
   - Cart Page: Destructures and passes ruleBasedQualifyingProductsMap
   - Add to Cart Modal: Uses qualifying products map for eligibility checks
   - Cart Product List: Accepts and forwards ruleBasedQualifyingProductsMap prop

6. Comprehensive Test Updates
   - Updated all test files to distinguish list-based vs rule-based promotions
   - Added new test cases for rule-based promotion eligibility scenarios
   - Enhanced test coverage across all bonus product utility test files

Backward Compatibility:
- All changes are backward compatible with existing list-based promotions
- Default parameter values ensure functions work without the new map
- Existing tests updated to explicitly mark promotions as list-based

* fix: prevent promotion message flashing when changing product variant (#3431)

* fix: prevent promotion flash by keeping previous data during variant changes

- Add keepPreviousData option parameter to useProductViewModal hook
- Pass keepPreviousData: true in BonusProductViewModal to maintain UI stability
- Prevents promotion message from clearing while fetching new variant data

* chore: trigger CI

* remove those stable references

* revert last changes

* bundle size limit update

* @W-20037720: Not allow duplicate items to appear in the cart (#3433)

* Fixed duplicating bonus products in the UI

* Fixed issue with list based not rendering properly

* Update cart-product-list-with-grouped-bonus-products.jsx

Signed-off-by: Daniel Diaz <38542490+ddiazccrz@users.noreply.github.com>

* Fixed lint errors

---------

Signed-off-by: Daniel Diaz <38542490+ddiazccrz@users.noreply.github.com>

---------

Signed-off-by: Daniel Diaz <38542490+ddiazccrz@users.noreply.github.com>
Co-authored-by: sf-xingquan-jin <xingquan.jin@salesforce.com>
Co-authored-by: cboscenco <cboscenco@salesforce.com>
Co-authored-by: Daniel Diaz <38542490+ddiazccrz@users.noreply.github.com>

ddiazccrz committed:
@W-20073841: Fixed intermittent issues with bonus product modal selection and add to cart (#3445)

* fixed issues on add to cart and showing bonus product modal

* Lint error

sf-shikhar-prasoon committed:
@W-20038640 Fix: 400 Error (#3443)

* fix

* add/update tests

* variant selection fix, test fix

* fix useVariant if product is already a variant

* only use react state for product-view modal

* removed the unnecessary useCallback

* add tests- Variation State Reset | URL Pollution Prevention

shethj committed:
Update changelogs

kevinxh reviewed (commented):
let's make sure 3.14 contains this fix - https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic/pull/248

let's coordinate with @joeluong-sfcc

shethj commented:
> let's make sure 3.14 contains this fix - [SalesforceCommerceCloud/commerce-sdk-isomorphic#248](https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic/pull/248)
> 
> let's coordinate with @joeluong-sfcc

Yeah I'm watching that PR. It requires to merge in and release the isomorphic-sdk.
I'll pull it in via version bump for the final release.

I'd like to get preview.0 out soon to allow smoke testing.
